### PR TITLE
Add new GraphQL types for budget investments

### PR DESCRIPTION
**Pull request by cyrillefr**
For now, in draft only status, as a start of a conversation.
## References

Attempt to close issue #4986 

## Objectives
Provide via GraphQL the like of https://demo.consuldemocracy.org/budgets/1/investments/16/ in JSON 
via the GraphQL API.

## Visual Changes

None. Technical changes only

## Notes

> Mention rake tasks or actions to be done when deploying this changes to a server (if any).
> Explain any caveats, or important things to notice like deprecations (if any).

### CODE REVIEW

**Comment by cyrillefr** [comment]
Hello @javierm ,

This is my first draft for this.
For being able to get some data from comments from investments, I had to modify the `comment` model,
regarding the `not_valuations` inside the `public_for_api` scope.
Some data are fetch, so technically, this is OK, though I am not sure it is correct from a business rule point of view.

Also, which rules should I test ?

**Comment by javierm** [comment]
Hi, @cyrillefr :smile:.

Thanks for the pull request! And sorry for the late reply :pray:.

> For being able to get some data from comments from investments, I had to modify the comment model,
regarding the not_valuations inside the public_for_api scope.

:+1:

> Also, which rules should I test ?

For now, you can check the tests in `lib/graphql_spec.rb` and write a new section `describe "Budget investments"` and add similar tests to what's there for (for instance) proposals.

**Comment by cyrillefr** [comment]
Hello @javierm,

I added the milestones type, so that it can be displayed alongside comments, like in the web version.

Please tell me if you think there are some missing fields.

**Review by javierm — COMMENTED** [review]
@cyrillefr Once again, thank you very much for this pull request! :pray:

I've left some comments. Let me know what you think! Feel free to disagree with some of them :wink:.

File: app/graphql/types/budget_type.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+module Types
+  class BudgetType < Types::BaseObject
+    field :id, ID, null: false
+    field :name, String, null: true
+    field :phase, String, null: true
+    field :investments, Types::InvestmentType, null: false do
+      argument :id, ID, required: true, default_value: false
```
[the remark this answers is not in the document]
Comment (cyrillefr) [reply app/graphql/types/budget_type.rb:8]: I have amended the code to do that.

File: app/graphql/types/investment_type.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+module Types
+  class InvestmentType < Types::BaseObject
+    field :id, ID, null: false
+    field :author, Types::UserType, null: true
```
[the remark this answers is not in the document]
Comment (cyrillefr) [reply app/graphql/types/investment_type.rb]: I have replaced `author` with `public_author`

File: app/graphql/types/investment_type.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    field :price, Integer, null: true
+    field :feasibility, String, null: true
+    field :tsv, String, null: true
```
[the remark this answers is not in the document]
Comment (cyrillefr) [reply app/graphql/types/investment_type.rb]: I just removed it.

File: app/graphql/types/investment_type.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
+module Types
+  class InvestmentType < Types::BaseObject
```
[the remark this answers is not in the document]
Comment (cyrillefr) [reply app/graphql/types/investment_type.rb]: I have chosen the first option, just like proposal_notifications

File: app/graphql/types/investment_type.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+    field :id, ID, null: false
+    field :author, Types::UserType, null: true
+    field :price, Integer, null: true
+    field :feasibility, String, null: true
+    field :tsv, String, null: true
+    field :title, String, null: true
+    field :description, String, null: true
+    field :location, String, null: true
+    field :comments, Types::CommentType.connection_type, null: true
```
[the remark this answers is not in the document]
Comment (cyrillefr) [reply app/graphql/types/investment_type.rb]: Yes, added :)

File: app/models/budget.rb [hunk 6]
```diff
@@ -58,6 +58,7 @@ def name_uniqueness_by_budget
   scope :balloting, -> { where(phase: "balloting") }
   scope :reviewing_ballots, -> { where(phase: "reviewing_ballots") }
   scope :finished, -> { where(phase: "finished") }
+  scope :public_for_api, -> { all }
```
[the remark this answers is not in the document]
Comment (cyrillefr) [reply app/models/budget.rb]: Replaced by `published`.

File: app/models/budget/investment.rb [hunk 7]
```diff
@@ -111,6 +111,7 @@ class Investment < ApplicationRecord
     end
 
     scope :for_render, -> { includes(:heading) }
+    scope :public_for_api, -> { all }
```
[the remark this answers is not in the document]
Comment (cyrillefr) [reply app/models/budget/investment.rb]: Yes, I changes the code.

File: app/graphql/types/investment_type.rb [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+    field :milestones, Types::MilestoneType.connection_type, null: true
```
[the remark this answers is not in the document]
Comment (cyrillefr) [reply app/graphql/types/investment_type.rb]: Added :)

File: app/models/milestone.rb [hunk 9]
```diff
@@ -16,6 +16,7 @@ class Milestone < ApplicationRecord
   scope :order_by_publication_date, -> { order(publication_date: :asc, created_at: :asc) }
   scope :published,                 -> { where("publication_date <= ?", Date.current.end_of_day) }
   scope :with_status,               -> { where.not(status_id: nil) }
+  scope :public_for_api, -> { all }
```
[the remark this answers is not in the document]
Comment (cyrillefr) [reply app/models/milestone.rb]: Added :)

File: spec/models/comment_spec.rb [hunk 10]
```diff
@@ -175,8 +175,15 @@
...
-    it "does not return comments on elements which are not debates or proposals" do
-      create(:comment, commentable: create(:budget_investment))
+    it "returns comments on budget investments" do
+      comment = create(:comment, commentable: create(:budget_investment))
+
+      expect(Comment.public_for_api).to eq [comment]
+    end
+
+    it "does not return comments on elements which are not debates, proposals
+      or budget investments" do
```
[the remark this answers is not in the document]
Comment (cyrillefr) [reply spec/models/comment_spec.rb]: I was afraid of the line length

**Comment by cyrillefr** [comment]
Hello @javierm , thanks for the tips.

**Review by javierm — APPROVED** [review]

**Review by javierm — APPROVED** [review]
@cyrillefr Thank you very much! :smile:

I've just rebased these changes against the latest master branch. We'll check whether we merge them before releasing version 2.2.1 or we leave them for version 2.3.

**Comment by javierm** [comment]
While reviewing this pull request, I noticed there was a small issue in our API code. I've just opened #5637 to solve it.

@cyrillefr During the reorganization of the tests for pull request #5637, I've taken the liberty to edit this pull request in order a few tests; hope you don't mind! :pray:

**Comment by cyrillefr** [comment]
Hello @javierm ,

No problemo.

**Review by taitus — APPROVED** [review]

**Comment by javierm** [comment]
@cyrillefr Once again, thank you very much for this pull request! :pray:

**Comment by cyrillefr** [comment]
Hello @javierm ,

Glad it works :)

**Pull request merged**